File: java_solutions/3363.find-the-maximum-number-of-fruits-collected.java
/*
 * @lc app=leetcode id=3363 lang=java
 *
 * [3363] Find the Maximum Number of Fruits Collected
 */

class Solution { public int maxCollectedFruits(int[][] grid) { return (int) ((Function<Integer, Integer>) n -> IntStream.range(0, n).map(i -> grid[i][i]).sum() + ((Function<Function<int[][], Integer>, Integer>) dp -> dp.apply(grid) + dp.apply(IntStream.range(0, n).mapToObj(i -> IntStream.range(0, n).map(j -> grid[j][i]).toArray()).toArray(int[][]::new))).apply((Function<int[][], Integer>) g -> IntStream.range(1, n - 1).boxed().reduce(IntStream.range(0, n).map(i -> i == n - 1 ? g[0][n - 1] : -1).toArray(), (prev, row) -> IntStream.range(0, n).map(i -> ((Function<Integer, Integer>) m -> m >= 0 ? m + g[row][i] : -1).apply(Math.max(i > 0 ? prev[i - 1] : -1, Math.max(prev[i], i < n - 1 ? prev[i + 1] : -1)))).toArray(), (a, b) -> a)[n - 1])).apply(grid.length); } }

File: kotlin_solutions_1/3363.find-the-maximum-number-of-fruits-collected.kt
/*
 * @lc app=leetcode id=3363 lang=java
 *
 * [3363] Find the Maximum Number of Fruits Collected
 */
class Solution {
    fun maxCollectedFruits(grid: Array<IntArray>): Int = grid.size.let { n -> (0 until n).sumOf { grid[it][it] } + arrayOf(true, false).sumOf { isTop -> Array(n) { IntArray(n) { -1 } }.apply { this[0][n - 1] = grid[0][n - 1] }.let { dp -> (1 until n - 1).forEach { i -> (maxOf(i + 1, n - 1 - i)..n - 1).forEach { j -> arrayOf(-1, 0, 1).map { dj -> if (isTop) dp[i - 1][j + dj] else dp[i - 1][j + dj] }.filter { it != -1 }.maxOrNull()?.let { prev -> dp[i][j] = prev + (if (isTop) grid[i][j] else grid[j][i]) } } }.run { dp[n - 2][n - 1] } } }
}